src/include/cpos.h

```diff
@@ -0,0 +1,485 @@
+// data CPO
+// @todo Figure out what is wrong with const
+// ----------------------------------------------------------------------------
+namespace _data {
+void data(auto&) = delete;
+void data(const auto&) = delete;
+
+template <class T>
+concept _member_data = requires(T t) {
+  { t.data() };
+};
+
+template <class T>
+concept _member_data_handle = requires(T t) {
+  { t.data_handle() };
+};
+
+struct _fn {
+  template <_member_data T>
+    requires(!_member_data_handle<T>)
+  constexpr auto operator()(T&& t) const noexcept {
+    return t.data();
+  }
+
+  template <_member_data_handle T>
+    requires(!_member_data<T>)
+  constexpr auto operator()(T&& t) const noexcept {
+    return t.data_handle();
+  }
+
+  template <_member_data_handle T>
+    requires(_member_data<T>)
+  constexpr auto operator()(T&& t) const noexcept {
+    return t.data_handle();
+  }
+};
+}  // namespace _data
+inline namespace _cpo {
+inline constexpr auto data = _data::_fn{};
+}  // namespace _cpo
+
+// ----------------------------------------------------------------------------
+// extents CPO
+// ----------------------------------------------------------------------------
+namespace _extents {
+void extents(auto&) = delete;
+void extents(const auto&) = delete;
+
+template <class T>
+concept _member_extents = requires(T t) {
+  { t.extents() };
+};
+
+template <typename T>
+concept _is_mdspan = std::same_as<
+    typename std::remove_cvref_t<T>,
+    stdx::mdspan<
+        typename std::remove_cvref_t<T>::value_type,
+        typename std::remove_cvref_t<T>::extents_type,
+        typename std::remove_cvref_t<T>::layout_type>>;
+
+struct _fn {
+  template <_member_extents T>
+    requires(!_is_mdspan<T>)
+  auto constexpr operator()(T&& t) const noexcept {
+    return t.extents();
+  }
+
+  template <_is_mdspan M>
+  auto constexpr operator()(M&& m) const noexcept {
+    return std::vector<size_t>{m.extents().extent(0), m.extents().extent(1)};
+  }
+};
+}  // namespace _extents
+inline namespace _cpo {
+inline constexpr auto extents = _extents::_fn{};
+}  // namespace _cpo
+
+// ----------------------------------------------------------------------------
+// load CPO
+// ----------------------------------------------------------------------------
+namespace _load {
+void load(auto&) = delete;
+void load(auto&&) = delete;
+void num_loads(auto&) = delete;
+void num_loads(auto&&) = delete;
+
+template <class T>
+concept _member_load = requires(T&& t) {
+  { t.load() } -> std::convertible_to<bool>;
+};
+
+template <class T>
+concept _member_num_loads = requires(T&& t) {
+  { t.num_loads() } -> semi_integral;
+};
+
+struct _fn {
+  template <_member_load T>
+  auto constexpr operator()(T&& t) const noexcept {
+    return t.load();
+  }
+
+  template <class T>
+    requires(!_member_load<T>)
+  auto constexpr operator()(T&& t) const noexcept {
+    // return t.load();
+    return false;
+  }
+};
```

Comment:
What's going on here? Should we remove `// return t.load();` because we're just using `requires(!_member_load<T>)`?